src/vs/editor/contrib/folding/foldingModel.ts

```diff
@@ -52,11 +56,28 @@ export class FoldingModel {
 			for (let region of regions) {
 				let index = region.regionIndex;
 				let editorDecorationId = this._editorDecorationIds[index];
+				let editorDecorationIdEnd = this._editorDecorationIdsEnd[index];
 				if (editorDecorationId && !processed[editorDecorationId]) {
 					processed[editorDecorationId] = true;
 					let newCollapseState = !this._regions.isCollapsed(index);
 					this._regions.setCollapsed(index, newCollapseState);
 					accessor.changeDecorationOptions(editorDecorationId, this._decorationProvider.getDecorationOption(newCollapseState));
+
+					if (editorDecorationIdEnd && !processed[editorDecorationIdEnd]) {
+						let regionsInside = this.getRegionsInside(region);
+
+						regionsInside.forEach(insideRegion => {
+							let insideRegionDecorationIdEnd = this._editorDecorationIdsEnd[insideRegion.regionIndex];
+
+							if (insideRegionDecorationIdEnd && !processed[insideRegionDecorationIdEnd]) {
+								processed[insideRegionDecorationIdEnd] = true;
+								accessor.changeDecorationOptions(insideRegionDecorationIdEnd, this._decorationProvider.getDecorationOption(newCollapseState, true, region.isCollapsed));
+							}
+						});
+
+						processed[editorDecorationIdEnd] = true;
+						accessor.changeDecorationOptions(editorDecorationIdEnd, this._decorationProvider.getDecorationOption(newCollapseState, true));
```

Comment:
What's the idea here?

Reply:
This one is the actual end decoration for the target region being toggled (versus the ones in the loop being hidden because they're inside the target region).